### PR TITLE
reimplementing __add__ method for spiketrains

### PR DESCRIPTION
**Pull request by grg2rsr**
for shifting all spikes in a spiketrain by a constant amount of time. Currently, when adding a time quantity to a spike train, several properties like `t_start`, `t_stop`, and most importantly, `annotations` are lost.

Hope I am doing everything right here, first pull request. Apologies if something is off normal procedure.

### CODE REVIEW

File: neo/core/spiketrain.py [hunk 1]
```diff
@@ -427,6 +427,23 @@ def __getslice__(self, i, j):
             obj.waveforms = obj.waveforms[i:j]
         return obj
 
+    def __add__(self,time):
```
Comment (apdavison) [inline neo/core/spiketrain.py]: add space after comma (PEP 8)

File: neo/core/spiketrain.py [hunk 2]
```diff
@@ -427,6 +427,23 @@ def __getslice__(self, i, j):
...
+        '''
+        Shifts the time point of all spikes by the amount in :attr:`time`.
+
+        Raises an exception of new time points fall outside :attr:`t_start` or
```
Comment (apdavison) [inline neo/core/spiketrain.py]: of --> if

File: neo/core/spiketrain.py [hunk 3]
```diff
@@ -427,6 +427,23 @@ def __getslice__(self, i, j):
             obj.waveforms = obj.waveforms[i:j]
         return obj
 
+    def __add__(self,time):
+        '''
+        Shifts the time point of all spikes by the amount in :attr:`time`.
+
```
Comment (apdavison) [inline neo/core/spiketrain.py]: You should document what type is expected for `time`, i.e. a scalar Quantity

File: neo/core/spiketrain.py [hunk 4]
```diff
@@ -427,6 +427,23 @@ def __getslice__(self, i, j):
...
+        Raises an exception of new time points fall outside :attr:`t_start` or
+        :attr:`t_stop`
+        '''
+        spikes = self.view(pq.Quantity)
+        check_has_dimensions_time(time)
+        _check_time_in_range(spikes+time, self.t_start, self.t_stop)
```
Comment (apdavison) [inline neo/core/spiketrain.py]: you should add space around operators ('+' in this case) - PEP 8

**Comment by grg2rsr** [comment]
thanks for the feedback, sorry for the bad docstring. Just a quick question, by adding tests I add functions to the`TestChanging` class in `test_spiketrain.py`?

**Comment by apdavison** [comment]
Yes, your tests are in the right place. Note that the tests are failing on Travis CI because of the way you are using `assertRaises`. Instead of:
```python
self.assertRaises(ValueError, train.__sub__(10 * pq.ms))
```
you need to use
```python
self.assertRaises(ValueError, train.__sub__, 10 * pq.ms)
```
or
```python
with self.assertRaises(ValueError):
    train.__sub__(10 * pq.ms)
```
see https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

**Comment by apdavison** [comment]
@grg2rsr I've now merged the PR. Thank-you again for the contribution.

For future reference, it is safer to make your changes in a branch other than master. This makes it easier to keep in sync with NeuralEnsemble/master in the case where the PR remains open for several days or weeks.

**Comment by grg2rsr** [comment]
thanks, I just saw it to and commited, I think you were faster though. Are the tests passing now? On my computer, they are not:
```
$ python test_spiketrain.py
...F......F...........................................E......................................
======================================================================
ERROR: test_duplicate_with_new_data (__main__.TestDuplicateWithNewData)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_spiketrain.py", line 1167, in test_duplicate_with_new_data
    signal1b = signal1.duplicate_with_new_data(new_data,
AttributeError: 'SpikeTrain' object has no attribute 'duplicate_with_new_data'

======================================================================
FAIL: test__adding_time (__main__.TestChanging)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_spiketrain.py", line 1366, in test__adding_time
    self.assertRaises(ValueError, train.__add__, 10 * pq.ms)
AssertionError: ValueError not raised

======================================================================
FAIL: test__subtracting_time (__main__.TestChanging)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_spiketrain.py", line 1373, in test__subtracting_time
    self.assertRaises(ValueError, train.__sub__, 10 * pq.ms)
AssertionError: ValueError not raised

----------------------------------------------------------------------
Ran 93 tests in 0.289s

FAILED (failures=2, errors=1)
```

**Comment by apdavison** [comment]
Yes, they are passing: https://travis-ci.org/NeuralEnsemble/python-neo/builds/197985499

**Comment by grg2rsr** [comment]
thanks you. 2 short questions:
1. do I need to set up any specific so my local tests match Travis CI tests (above output was generated by following the instructions in the `neo/test/README.txt` and they still fail, even with the now corrected syntax
2. For the future, what is the correct approach for submitting a pull request, to which branch should I compare mine to?

**Pull request merged**